File: src/main/java/com/growingio/kafka/TimestampFuzzyComparator.java
package com.growingio.kafka;

/**
 * Created by foolchi on 01/08/15.
 */
public class TimestampFuzzyComparator implements FuzzyBinaryComparator {
    private long maxDiff;

    public TimestampFuzzyComparator(long maxDiff) {
        this.maxDiff = maxDiff;
    }

    public int exactCompare(String msg, String dest) {
        long diff = Long.parseLong(msg) - Long.parseLong(dest);
        if (diff > 0)
            return 1;
        if (diff < 0)
            return -1;
        return 0;
    }

    public int compare(String msg, String dest) {
        long diff = Long.parseLong(msg) - Long.parseLong(dest);
        if (Math.abs(diff) < maxDiff) {
            return 0;
        }
        if (diff < 0)
            return -1;
        return 1;
    }
}

File: src/main/scala/com/growingio/kafka/TimestampFuzzyComparator.scala
package com.growingio.kafka

/**
 * Created by foolchi on 03/08/15.
 * 带模糊匹配的时间戳比较类
 * maxDiff是默认的最大误差(+/-)
 */
class TimestampFuzzyComparator (dest : String, val maxDiff : Long)extends FuzzyBinaryComparator{
  private val destTime = dest.toLong

  override def exactCompare(msg: String): Int = {
    val diff = msg.toLong - destTime
    if (diff > 0) 1 else if (diff < 0) -1 else 0
  }

  override def compare(msg: String): Int = {
    val diff = msg.toLong - destTime
    if (diff.abs < maxDiff) 0 else if (diff < 0) -1 else 1
  }
}
